src/runtime_src/core/tools/xbutil2/OO_P2P.cpp

```diff
@@ -17,12 +17,244 @@
+  long long exp_bar = -1;
+
+  for (auto& str : cfg) {
+    auto pos = str.find(":");
+    auto key = str.substr(0, pos);
+    ++pos; // past ":"
+    if (key == "bar")
+      bar = std::stoll(str.substr(pos));
+    else if (key == "exp_bar")
+      exp_bar = std::stoll(str.substr(pos));
+    else if (key == "rbar")
+      rbar = std::stoll(str.substr(pos));
+    else if (key == "remap")
+      remap = std::stoll(str.substr(pos));
+  }
+
+  if (bar == -1)
+    throw xrt_core::system_error(ENOTSUP,"p2p is not supported");
+  if (rbar != -1 && rbar > bar)
+    throw xrt_core::system_error(EIO,"Please WARM reboot to enable p2p");
+  if (remap > 0 && remap != bar)
+    throw xrt_core::system_error(EIO,"p2p remapper is not set correctly");
+  if (bar == exp_bar)
+    return;
+
+  throw xrt_core::system_error(ENOTSUP,"p2p is not supported");
+}
+
+namespace p2ptest {
+
+static void
+fill_with_stride(char* begin, const char* end, char pattern)
+{
+  auto stride = xrt_core::getpagesize();
+  if (std::distance(static_cast<const char*>(begin), end) % stride)
+    throw xrt_core::system_error(EINVAL, "Range not an increment of stride: " + std::to_string(stride));
+  for (auto itr = begin; itr != end; itr += stride)
+    (*itr) = pattern;
+}
+
+static void
+cmp_with_stride(const char* begin, const char* end, char pattern)
+{
+  auto stride = xrt_core::getpagesize();
+  if (std::distance(begin, end) % stride)
+    throw xrt_core::system_error(EINVAL, "Range not an increment of stride: " + std::to_string(stride));
+  for (auto itr = begin; itr != end; itr += stride) {
+    if ((*itr) != pattern) {
+      std::ostringstream ostr;
+      ostr << "Error in p2p comparison, expected '"
+           << pattern << "' got '" << (*itr) << "'";
+      throw xrt_core::system_error(EIO, ostr.str());
+    }
+  }
+}
+
+static void
+chunk(xrt_core::device* device, char* boptr, uint64_t dev_addr, uint64_t size)
+{
+  char patternA = 'A';
+  char patternB = 'B';
+
+  auto mem = xrt_core::aligned_alloc(xrt_core::getpagesize(), size);
+  auto buf = static_cast<char*>(mem.get());
+
+  fill_with_stride(buf, buf+size, patternA);
+  try {
+    device->unmgd_pwrite(buf, size, dev_addr);
+  }
+  catch (const std::exception&) {
+    std::ostringstream ostr;
+    ostr << "Error writing 0x" << std::hex << size
+         << " bytes to 0x" << dev_addr << std::dec;
+    throw xrt_core::system_error(EIO, ostr.str());
+  }
+  cmp_with_stride(boptr, boptr+size, patternA);
+
+  fill_with_stride(boptr, boptr+size, patternB);
+  try {
+    device->unmgd_pread(buf, size, dev_addr);
+  }
+  catch (const std::exception&) {
+    std::ostringstream ostr;
+    ostr << "Error reading 0x" << std::hex << size
+         << " bytes from 0x" << dev_addr << std::dec;
+    throw xrt_core::system_error(EIO, ostr.str());
+  }
+  cmp_with_stride(buf, buf+size, patternB);
+}
+
+static void
+bank(xrt_core::device* device, int memidx, uint64_t addr, uint64_t size)
+{
+  const size_t chunk_size = 16 * 1024 * 1024;
```

Comment:
Can the magic number "16" be assigned to a `static const size_t `variable, especially since it is used more then once.

Reply:
I think a comment will do explaining what the function does.  I pretty much just copied it and cleaned it up a litte